src/Propulsion/Streams.fs

```diff
@@ -84,57 +84,78 @@ module StreamSpan =
-            res
-        let trimmed = span |> Array.takeWhile withinLimits
-        metrics eventSize trimmed, trimmed
-
-    let inline idx (span: FsCodec.ITimelineEvent<'F>[]) = span[0].Index
-    let inline ver (span: FsCodec.ITimelineEvent<'F>[]) = idx span + span.LongLength
-    let dropBeforeIndex min: FsCodec.ITimelineEvent<_>[] -> FsCodec.ITimelineEvent<_>[] = function
-        | xs when xs.Length = 0 -> null
-        | xs when idx xs >= min -> xs // don't adjust if min not within
-        | v when ver v <= min -> null // throw away if before min
-        | xs -> xs |> Array.skip (min - idx xs |> int) // slice
+            let fitsAndNotAnUnfold = (countBudget >= 0 && bytesBudget >= 0) && not x.IsUnfold
+            if fitsAndNotAnUnfold then count <- count + 1; bytes <- bytes + eventBytes
+            fitsAndNotAnUnfold
+        let trimmedEvents = span |> Array.takeWhile withinLimits
+        // takeWhile terminated either because it hit the first Unfold, or the size limit
+        // In either case, if the next event is an Unfold, we know it (and any successors) must be associated with that final event
+        if span |> Array.tryItem trimmedEvents.Length |> Option.exists _.IsUnfold then span
+        else trimmedEvents
+    let slice<'F> eventSize limits (span: FsCodec.ITimelineEvent<'F>[]): struct (FsCodec.ITimelineEvent<'F>[] * Metrics) =
+        let trimmed =
+            // we must always send one event, even if it exceeds the limit (if the handler throws, the the Stats can categorize the problem to surface it)
+            if span[0].IsUnfold || span.Length = 1 || span[1].IsUnfold then span
+            // If we have 2 or more (non-Unfold) events, then we limit the batch size
+            else trimEvents<'F> eventSize limits span
+        trimmed, metrics eventSize trimmed
+
+    let inline index (span: FsCodec.ITimelineEvent<'F>[]) = span[0].Index
+    let inline nextIndex (span: FsCodec.ITimelineEvent<'F>[]) =
+        let l = span[span.Length - 1]
+        if l.IsUnfold then l.Index else l.Index + 1L
+    let inline dropBeforeIndex min = function
+        | [||] as xs -> xs
+        | xs when nextIndex xs < min -> Array.empty
+        | xs ->
+            match index xs with
+            | xi when xi = min -> xs
+            | xi -> xs |> Array.skip (min - xi |> int)
 
     let merge min (spans: FsCodec.ITimelineEvent<_>[][]) =
-        let candidates = [|
-            for span in spans do
-                if span <> null then
-                    match dropBeforeIndex min span with
-                    | null -> ()
-                    | trimmed when trimmed.Length = 0 -> invalidOp "Cant add empty"
-                    | trimmed -> trimmed |]
+        let candidates = [| for span in spans do
+                                if span <> null then
+                                    match dropBeforeIndex min span with
+                                    | [||] -> ()
+                                    | xs -> xs |]
```

Comment:
was better before the added indentation 😓

Reply:
it was an improvement at one point in the refactoring cycle, I promise, but agreed...